### PR TITLE
#263: extract method and make cacheable

### PR DESCRIPTION
**Pull request by lautarobock**
#263 Extract map method creator and make it cacheable.

### CODE REVIEW

**Comment by davvd** [comment]
@lautarobock Thanks, I will ask someone to do a code review here, thanks for your pull request

**Comment by davvd** [comment]
@carlosmiranda please review, thanks

**Comment by lautarobock** [comment]
@davvd @carlosmiranda any news here?

File: src/main/java/org/takes/rq/RqForm.java [hunk 1]
```diff
@@ -160,6 +154,44 @@ private static String decode(final CharSequence txt) {
...
+         */
+        private ConcurrentMap<String, List<String>> map() throws IOException  {
+            synchronized (this.saved) {
+                if (this.saved.isEmpty()) {
+                    final ByteArrayOutputStream
+                        baos = new ByteArrayOutputStream();
+                    new RqPrint(this.req).printBody(baos);
+                    final String body = new String(baos.toByteArray());
+                    final ConcurrentMap<String, List<String>> map =
+                        new ConcurrentHashMap<String, List<String>>(0);
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqForm.java]: @lautarobock Better to use an intial size of `1`.

File: src/main/java/org/takes/rq/RqForm.java [hunk 2]
```diff
@@ -183,11 +215,13 @@ public Smart(final RqForm req) {
             this.origin = req;
         }
         @Override
-        public Iterable<String> param(final CharSequence name) {
+        public Iterable<String> param(
+            final CharSequence name
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqForm.java]: @lautarobock This param can be in the line above (you can declare `throws` in this line instead).

File: src/main/java/org/takes/rq/RqForm.java [hunk 3]
```diff
@@ -221,8 +255,11 @@ public String single(final CharSequence name) throws IOException {
          * @param name Name of query param
          * @param def Default, if not found
          * @return Value of it
+         * @throws IOException if something fails reading parameters
          */
-        public String single(final CharSequence name, final String def) {
+        public String single(
+                final CharSequence name, final String def
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqForm.java]: @lautarobock Same here, you can keep these params in the previous line and declare the throws clause in this line.

File: src/main/java/org/takes/rq/RqForm.java [hunk 4]
```diff
@@ -68,67 +69,60 @@
...
+        private final transient List<ConcurrentMap<String, List<String>>> saved =
+            new CopyOnWriteArrayList<ConcurrentMap<String, List<String>>>();
+
         /**
          * Ctor.
-         * @param req Original request
+         * @param request Original request
          * @throws IOException If fails
          */
         @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
```
Comment (carlosmiranda) [inline src/main/java/org/takes/rq/RqForm.java]: @lautarobock We don't need this warning here anymore.

**Comment by carlosmiranda** [comment]
@lautarobock Apologies for the delay. Please see my comments above.

**Comment by lautarobock** [comment]
@carlosmiranda done, please check the update

**Comment by carlosmiranda** [comment]
@rultor good to merge

**Comment by yegor256** [comment]
@lautarobock no tests? how do you know it works?

**Comment by lautarobock** [comment]
@yegor256 the test was there and I've made only a inner refactor. But I've added a test for checking cache.

**Comment by lautarobock** [comment]
@carlosmiranda @yegor256 Anybody saw the last commit?

**Comment by yegor256** [comment]
@rultor merge

**Comment by davvd** [comment]
@elenavolokhova please, review this task for compliance with our [quality rules](http://www.teamed.io/qa.html)

**Comment by davvd** [comment]
@rultor please deploy

**Comment by elenavolokhova** [comment]
@carlosmiranda 
According to our [quality rules](http://www.teamed.io/qa.html):
 > Comments were mostly about design problems, not cosmetic issues.
 
 There were mostly cosmetic issues found during review. Try to pay attention to bigger design problems and the overall quality of the code.
Please confirm that this new rule is clear.

**Comment by carlosmiranda** [comment]
@elenavolokhova All is clear, thanks.

**Comment by elenavolokhova** [comment]
@davvd Quality is acceptable here.

**Comment by davvd** [comment]
> @davvd Quality is acceptable here.

@elenavolokhova thanks a lot, next time everybody should try to make it better

**Comment by davvd** [comment]
@carlosmiranda just added 10 mins to @elenavolokhova (for QA), payment ID is `59513908`... Much appreciated! **18 mins** was added to your account, payment ID `59513917`, time spent is 166 hours and 44 mins... you're getting extra minutes here (c=3)... added +18 to your rating, now it is equal to [+4134](http://www.netbout.com/b/35023?open=rating)

**Pull request merged**